### DISCUSSION TITLE
聚合查询耗时长

### DISCUSSION THREAD

**Discussion by michael1991** [apache/doris#19780] [A - General / Q&A]
Aggregate模型，数据量在700w左右，简单的聚合查询（聚合字段非key）耗时在500-600ms，查看profile主要耗时在DataArrivalWaitTime，FirstBatchArrivalWaitTime，请问这意味着什么？该如何优化呢？

**Comment by yagagagaga** [comment]
DataArrivalWaitTime表示数据到达前到等待时间，FirstBatchArrivalWaitTime表示第一批数据到达前到等待时间，通常这两个都是在VEXCHANGE_NODE节点下（即shuffle阶段发生的），对于聚合查询（聚合字段非key），会花比较多的时间做shuffle，这是正常现象，如果需要知道哪个点可以优化，还需要贴出完整的profile和告知Doris版本。

**Reply by michael1991** [reply]
大概明白了，那请问如果将聚合字段改为key column/partition column，是否可以避免shuffle？

**Reply by michael1991** [reply]
做了些调整，把聚合字段设置为key并做了分区，似乎更慢了
**Doris Version: 1.2.4.1**
单节点部署

Table: 
```sql
CREATE TABLE IF NOT EXISTS test.table
(
    `datehour` DATETIMEV2 NOT NULL COMMENT "时间",
    `id` VARCHAR(60) NOT NULL COMMENT "ID",
    `flag` TINYINT MAX DEFAULT "0" COMMENT "标识",
    `cost` DOUBLE MIN COMMENT "COST"
)
AGGREGATE KEY(`datehour`, `id`)
PARTITION BY RANGE(`datehour`) ()
DISTRIBUTED BY HASH(`id`)
PROPERTIES (
    "replication_num" = "1",
    "dynamic_partition.enable" = "true",
    "dynamic_partition.time_unit" = "HOUR",
    "dynamic_partition.start" = "-72",
    "dynamic_partition.end" = "1",
    "dynamic_partition.prefix" = "p",
    "dynamic_partition.buckets" = "2",
    "dynamic_partition.storage_medium" = "SSD",
    "dynamic_partition.create_history_partition" = "true",
    "dynamic_partition.history_partition_num" = "72"
); 
```
SQL: 
```sql
select datehour,sum(cost) from test.table where flag=1 group by datehour order by datehour;
```

**Fragments**
         ┌───────────────────────┐
         │[-1: VDataBufferSender]│
         │Fragment: 0            │
         │MaxActiveTime: 1s100ms │
         └───────────────────────┘
                     │
                     │
           ┌───────────────────┐
           │[5: VEXCHANGE_NODE]│
           │Fragment: 0        │
           └───────────────────┘
                     └┐
                      │
          ┌──────────────────────┐
          │[5: VDataStreamSender]│
          │Fragment: 1           │
          │MaxActiveTime: 1s100ms│
          └──────────────────────┘
                     ┌┘
                     │
             ┌───────────────┐
             │[2: VSORT_NODE]│
             │Fragment: 1    │
             └───────────────┘
                     └┐
                      │
          ┌──────────────────────┐
          │[4: VAGGREGATION_NODE]│
          │Fragment: 1           │
          └──────────────────────┘
                     ┌┘
                     │
           ┌───────────────────┐
           │[3: VEXCHANGE_NODE]│
           │Fragment: 1        │
           └───────────────────┘
                     └┐
                      │
          ┌──────────────────────┐
          │[3: VDataStreamSender]│
          │Fragment: 2           │
          │MaxActiveTime: 1s100ms│
          └──────────────────────┘
                      │
                      │
          ┌──────────────────────┐
          │[1: VAGGREGATION_NODE]│
          │Fragment: 2           │
          └──────────────────────┘
                      │
                      │
┌──────────────────────────────────────────┐
│[0: VNewOlapScanNode(log_details_rt_base)]│
│Fragment: 2                               │
└──────────────────────────────────────────┘
                     ┌┘
                     │
               ┌───────────┐
               │[VScanner] │
               │Fragment: 2│
               └───────────┘
                     │
                     │
            ┌─────────────────┐
            │[SegmentIterator]│
            │Fragment: 2      │
            └─────────────────┘

**Instance-0**
    ┌──────────────────────────────────┐
    │[-1: VDataBufferSender]           │
    │(Active: 0ns, non-child: 0.00)    │
    │  - Counters:                     │
    │      - AppendBatchTime: 142.180us│
    │      - NumSentRows: 25           │
    └──────────────────────────────────┘
                      │
                      │
┌──────────────────────────────────────────┐
│[5: VEXCHANGE_NODE]                       │
│(Active: 1s100ms, non-child: 97.50)       │
│  - Counters:                             │
│      - BytesReceived: 0.00               │
│      - DataArrivalWaitTime: 1s100ms      │
│      - DecompressBytes: 0.00             │
│      - DecompressTime: 0ns               │
│      - DeserializeRowBatchTimer: 0ns     │
│      - FirstBatchArrivalWaitTime: 1s100ms│
│      - MergeGetNext: 84.370us            │
│      - MergeGetNextBlock: 83.650us       │
│      - PeakMemoryUsage: 1.92 KB          │
│      - ProjectionTime: 0ns               │
│      - RowsReturned: 25                  │
│      - RowsReturnedRate: 22              │
│      - SendersBlockedTotalTimer(*): 0ns  │
└──────────────────────────────────────────┘

**Instance-2**
                                                                                                                                                                                                                                                                                                                    ┌──────────────────────────────────────────────┐
                                                                                                                                                                                                                                                                                                                    │[3: VDataStreamSender]                        │
                                                                                                                                                                                                                                                                                                                    │(Active: 41.640us, non-child: 0.00)           │
                                                                                                                                                                                                                                                                                                                    │  - Counters:                                 │
                                                                                                                                                                                                                                                                                                                    │      - BlocksSent: 2                         │
                                                                                                                                                                                                                                                                                                                    │      - BrpcSendTime: 90.900us                │
                                                                                                                                                                                                                                                                                                                    │      - BrpcSendTime.Wait: 181.710us          │
                                                                                                                                                                                                                                                                                                                    │      - BytesSent: 0.00                       │
                                                                                                                                                                                                                                                                                                                    │      - CompressTime: 0ns                     │
                                                                                                                                                                                                                                                                                                                    │      - IgnoreRows: 0                         │
                                                                                                                                                                                                                                                                                                                    │      - LocalBytesSent: 1.46 KB               │
                                                                                                                                                                                                                                                                                                                    │      - LocalSendTime: 28.360us               │
                                                                                                                                                                                                                                                                                                                    │      - LocalSentRows: 25                     │
                                                                                                                                                                                                                                                                                                                    │      - OverallThroughput: 0.0 /sec           │
                                                                                                                                                                                                                                                                                                                    │      - PeakMemoryUsage: 3.19 KB              │
                                                                                                                                                                                                                                                                                                                    │      - SerializeBatchTime: 0ns               │
                                                                                                                                                                                                                                                                                                                    │      - SplitBlockDistributeByChannelTime: 0ns│
                                                                                                                                                                                                                                                                                                                    │      - SplitBlockHashComputeTime: 0ns        │
                                                                                                                                                                                                                                                                                                                    │      - UncompressedRowBatchSize: 0.00        │
                                                                                                                                                                                                                                                                                                                    └──────────────────────────────────────────────┘
                                                                                                                                                                                                                                                                                                                                            │
                                                                                                                                                                                                                                                                                                                                            │
                                                                                                                                                                                                                                                                                                                   ┌────────────────────────────────────────────────┐
                                                                                                                                                                                                                                                                                                                   │[1: VAGGREGATION_NODE]                          │
                                                                                                                                                                                                                                                                                                                   │(Active: 1s97ms, non-child: 64.12)              │
                                                                                                                                                                                                                                                                                                                   │  - Info:                                       │
                                                                                                                                                                                                                                                                                                                   │      - ExecOption: Streaming Preaggregation    │
                                                                                                                                                                                                                                                                                                                   │  - Counters:                                   │
                                                                                                                                                                                                                                                                                                                   │      - BuildTime: 717.808ms                    │
                                                                                                                                                                                                                                                                                                                   │      - DeserializeDataTime: 0ns                │
                                                                                                                                                                                                                                                                                                                   │      - ExecTime: 510.209ms                     │
                                                                                                                                                                                                                                                                                                                   │      - ExprTime: 1.357ms                       │
                                                                                                                                                                                                                                                                                                                   │      - GetResultsTime: 33.460us                │
                                                                                                                                                                                                                                                                                                                   │      - HashTableComputeTime: 204.366ms         │
                                                                                                                                                                                                                                                                                                                   │      - HashTableInputCount: 7.077334M (7077334)│
                                                                                                                                                                                                                                                                                                                   │      - HashTableIterateTime: 610ns             │
                                                                                                                                                                                                                                                                                                                   │      - HashTableSize: 0                        │
                                                                                                                                                                                                                                                                                                                   │      - InsertKeysToColumnTime: 1.600us         │
                                                                                                                                                                                                                                                                                                                   │      - MaxRowSizeInBytes: 0                    │
                                                                                                                                                                                                                                                                                                                   │      - MergeTime: 0ns                          │
                                                                                                                                                                                                                                                                                                                   │      - PeakMemoryUsage: 0.00                   │
                                                                                                                                                                                                                                                                                                                   │      - ProjectionTime: 0ns                     │
                                                                                                                                                                                                                                                                                                                   │      - RowsReturned: 25                        │
                                                                                                                                                                                                                                                                                                                   │      - RowsReturnedRate: 22                    │
                                                                                                                                                                                                                                                                                                                   │      - SerializeDataTime: 21.660us             │
                                                                                                                                                                                                                                                                                                                   │      - SerializeKeyTime: 0ns                   │
                                                                                                                                                                                                                                                                                                                   │      - SerializeResultTime: 40.760us           │
                                                                                                                                                                                                                                                                                                                   │      - StreamingAggTime: 0ns                   │
                                                                                                                                                                                                                                                                                                                   └────────────────────────────────────────────────┘
                                                                                                                                                                                                                                                                                                                                            │
                                                                                                                                                                                                                                                                                                                                            │
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│[0: VNewOlapScanNode(log_details_rt_base)]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             │
│(Active: 374.1ms, non-child: 33.14)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    │
│  - Info:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              │
│      - RemainedDownPredicates: VectorizedFn[eq({"flag", "Int8"})]{                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                │
│SlotRef(slot_id=2 type=TINYINT codegen=false),                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         │
│VLiteral (name = Int8, type = Int8, value = (-1971145825))}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           │
│      - KeyRanges: ScanKeys:ScanKey=[null(0000-01-01 00:00:00) : 9999-12-31 23:59:59.999999]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           │
│      - TabletIds: [11131, 11143, 11170, 11013, 11128, 11156, 11116, 11111, 11153, 11123, 11106, 11151, 11096, 11148, 11023, 11167, 11061, 11113, 11093, 10966, 10996, 11164, 10986, 10988, 11026, 10983, 11011, 11158, 11118, 11194, 10963, 11066, 10968, 11078, 10998, 11101, 10978, 10991, 11138, 10981, 11043, 11103, 11021, 11083, 11141, 11041, 10993, 11179, 11036, 11003, 11038, 11058, 11081, 11088, 11126, 11016, 11006, 10971, 11108, 11053, 11071, 11056, 11068, 11001, 11046, 11076, 11188, 11051, 11091, 11018, 11048, 11031, 11133, 11033, 10973, 10961, 11173, 11191, 11008, 11063, 11086, 11121, 11176, 11185, 11098, 10976, 11161, 11136, 11073, 11146, 11182, 11028]│
│      - PushDownPredicates: [{flag IN [1]}]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        │
│  - Counters:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          │
│      - AcuireRuntimeFilterTime: 530ns                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 │
│      - GetNextTime: 353.12ms                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          │
│      - MaxScannerThreadNum: 12                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        │
│      - NewlyCreateFreeBlocksNum: 229                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  │
│      - NumDiskAccess: 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               │
│      - NumScanners: 92                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                │
│      - PeakMemoryUsage: 0.00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          │
│      - PreAllocFreeBlocksNum: 60                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      │
│      - ProcessConjunctTime: 47.290us                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  │
│      - ProjectionTime: 0ns                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            │
│      - RowsRead: 7.262876M (7262876)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  │
│      - RowsReturned: 7.077334M (7077334)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              │
│      - RowsReturnedRate: 18.923298M /sec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              │
│      - ScannerBatchWaitTime: 320.534ms                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                │
│      - ScannerCtxSchedCount: 281                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      │
│      - ScannerSchedCount: 281                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         │
│      - ScannerWorkerWaitTime: 777.18ms                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                │
│      - TabletNum: 92                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  │
│      - TotalReadThroughput: 0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
                                                                                                                                                                                                                                                                                                                                           ┌┘
                                                                                                                                                                                                                                                                                                                                           │
                                                                                                                                                                                                                                                                                                                        ┌─────────────────────────────────────┐
                                                                                                                                                                                                                                                                                                                        │[VScanner]                           │
                                                                                                                                                                                                                                                                                                                        │(Active: 0ns, non-child: 0.00)       │
                                                                                                                                                                                                                                                                                                                        │  - Counters:                        │
                                                                                                                                                                                                                                                                                                                        │      - BlockConvertTime: 0ns        │
                                                                                                                                                                                                                                                                                                                        │      - BlockFetchTime: 1s733ms      │
                                                                                                                                                                                                                                                                                                                        │      - ReaderInitTime: 0ns          │
                                                                                                                                                                                                                                                                                                                        │      - RowsDelFiltered: 0           │
                                                                                                                                                                                                                                                                                                                        │      - ScannerConvertBlockTime: 0ns │
                                                                                                                                                                                                                                                                                                                        │      - ScannerCpuTime: 3s746ms      │
                                                                                                                                                                                                                                                                                                                        │      - ScannerFilterTime: 591.538ms │
                                                                                                                                                                                                                                                                                                                        │      - ScannerGetBlockTime: 12s107ms│
                                                                                                                                                                                                                                                                                                                        │      - ScannerInitTime: 3.94ms      │
                                                                                                                                                                                                                                                                                                                        │      - ScannerPrefilterTime: 0ns    │
                                                                                                                                                                                                                                                                                                                        └─────────────────────────────────────┘
                                                                                                                                                                                                                                                                                                                                           └┐
                                                                                                                                                                                                                                                                                                                                            │
                                                                                                                                                                                                                                                                                                                    ┌──────────────────────────────────────────────┐
                                                                                                                                                                                                                                                                                                                    │[SegmentIterator]                             │
                                                                                                                                                                                                                                                                                                                    │(Active: 0ns, non-child: 0.00)                │
                                                                                                                                                                                                                                                                                                                    │  - Counters:                                 │
                                                                                                                                                                                                                                                                                                                    │      - BitmapIndexFilterTimer: 42.939us      │
                                                                                                                                                                                                                                                                                                                    │      - BlockConditionsFilteredTime: 118.235us│
                                                                                                                                                                                                                                                                                                                    │      - BlockInitSeekCount: 2.163K (2163)     │
                                                                                                                                                                                                                                                                                                                    │      - BlockInitSeekTime: 9.588ms            │
                                                                                                                                                                                                                                                                                                                    │      - BlockInitTime: 94.699ms               │
                                                                                                                                                                                                                                                                                                                    │      - BlockLoadTime: 1s732ms                │
                                                                                                                                                                                                                                                                                                                    │      - BlocksLoad: 2.731K (2731)             │
                                                                                                                                                                                                                                                                                                                    │      - CachedPagesNum: 0                     │
                                                                                                                                                                                                                                                                                                                    │      - CompressedBytesRead: 322.66 MB        │
                                                                                                                                                                                                                                                                                                                    │      - DecompressorTimer: 487.901ms          │
                                                                                                                                                                                                                                                                                                                    │      - FirstReadSeekCount: 381               │
                                                                                                                                                                                                                                                                                                                    │      - FirstReadSeekTime: 198.599ms          │
                                                                                                                                                                                                                                                                                                                    │      - FirstReadTime: 1s615ms                │
                                                                                                                                                                                                                                                                                                                    │      - IOTimer: 223.67ms                     │
                                                                                                                                                                                                                                                                                                                    │      - LazyReadSeekCount: 0                  │
                                                                                                                                                                                                                                                                                                                    │      - LazyReadSeekTime: 0ns                 │
                                                                                                                                                                                                                                                                                                                    │      - LazyReadTime: 0ns                     │
                                                                                                                                                                                                                                                                                                                    │      - NumSegmentFiltered: 0                 │
                                                                                                                                                                                                                                                                                                                    │      - NumSegmentTotal: 381                  │
                                                                                                                                                                                                                                                                                                                    │      - OutputColumnTime: 685.520us           │
                                                                                                                                                                                                                                                                                                                    │      - RawRowsRead: 7.443006M (7443006)      │
                                                                                                                                                                                                                                                                                                                    │      - RowsBitmapIndexFiltered: 0            │
                                                                                                                                                                                                                                                                                                                    │      - RowsBloomFilterFiltered: 0            │
                                                                                                                                                                                                                                                                                                                    │      - RowsConditionsFiltered: 0             │
                                                                                                                                                                                                                                                                                                                    │      - RowsKeyRangeFiltered: 0               │
                                                                                                                                                                                                                                                                                                                    │      - RowsShortCircuitPredFiltered: 0       │
                                                                                                                                                                                                                                                                                                                    │      - RowsShortCircuitPredInput: 0          │
                                                                                                                                                                                                                                                                                                                    │      - RowsStatsFiltered: 0                  │
                                                                                                                                                                                                                                                                                                                    │      - RowsVectorPredFiltered: 0             │
                                                                                                                                                                                                                                                                                                                    │      - RowsVectorPredInput: 0                │
                                                                                                                                                                                                                                                                                                                    │      - ShortPredEvalTime: 0ns                │
                                                                                                                                                                                                                                                                                                                    │      - TotalPagesNum: 10.735K (10735)        │
                                                                                                                                                                                                                                                                                                                    │      - UncompressedBytesRead: 423.70 MB      │
                                                                                                                                                                                                                                                                                                                    │      - VectorPredEvalTime: 0ns               │
                                                                                                                                                                                                                                                                                                                    └──────────────────────────────────────────────┘

**Reply by ZhFangyuan** [reply]
你好，请问这个问题你解决了吗

**Reply by michael1991** [reply]
我们最后牺牲了一些实时性，来缩短查询耗时，在Aggregate模型上定时构建Duplicate模型，然后查询Duplicate模型；
直接在Aggregate模型上聚合慢的原因也在上面解释了哈

**Reply by ZhFangyuan** [reply]
谢谢回复~~
